### ISSUE TITLE
Markdown and MathJax

### ISSUE THREAD

**Issue by samfearn** [maths/moodle-qtype_stack#420]
Hi,

I can't tell whether this is a stack problem, or a problem with how I have moodle set up, so apologies in advance.

I like to write question text using Markdown, and so I have my default moodle editor as the plain text area and use the markdown 'type' dropdown option. In other question types (non-stack) I can then include LaTeX between $ or $$'s depending on whether I want inline or display math. I know that stack doesn't want me to use dollars, so I should use \( and \) for inline and \[ and \] for display math. This works if I set the plain text 'type' to be html, so far so good. Let me use the following question as an example; <p> tags are missing but it displays ok aside from that:

```
Let \(f(x)\) be defined as \[f(x)=x^2+1\]

What is the value of \( f\left(\frac{1}{2} \right) \)?

[[input:ans1]] [[validation:ans1]]
```

If I take the same question and set the text type to markdown, then it doesn't get processed by MathJax. My first thought was that this was an issue of markdown itself removing the backslashes, and so I tried escaping the backslashes and still get no processing. In fact, using a web console I can see that MathJax.js never even gets loaded. However, if I now add $x$ at the start of the first line, as well as escaping my backslashes,

```
$x$Let \\(f(x)\\) be defined as \\[f(x)=x^2+1\\]

What is the value of \\( f\left(\frac{1}{2} \right) \\)?

[[input:ans1]] [[validation:ans1]]
```

then suddenly everything works as expected -- MathJax.js is loaded and everything is formatted like I'd expect. So this seems to be an issue of MathJax not getting loaded properly.

I don't know whether this is a stack issue, or moodle issue, or a problem with my MathJax settings, so FWIW, in Admin->Plugins->MathJax I have MathJax configuration set as:

```
MathJax.Hub.Config({
    tex2jax: {
    inlineMath: [['$','$'], ['\\(','\\)']],
    processEscapes: true},
    config: ["Accessible.js", "Safe.js"],
    extensions: ["tex2jax.js"],
    errorSettings: { message: ["!"] },
    skipStartupTypeset: true,
    messageStyle: "none"
});
```

and the additional equation delimiters set as `$,\(,\),\\(,\\),\[,\]`.

**Comment by sangwinc** [comment]
Sorry about this.

I can reproduce the issue, exactly as you describe.  Markdown is certainly a format we want to support, so we do need to fix this.

The backslashes are being stripped out here: https://github.com/maths/moodle-qtype_stack/blob/master/renderer.php#L81
(but not by this: https://github.com/maths/moodle-qtype_stack/blob/master/renderer.php#L82)

The same problem occurs with the "plain text" format as well as the "markdown" format.

It looks like the backslashes are getting stripped out before the filter is applied to create the <span class="filter_mathjaxloader_equation">Let <span class="nolink">\(f(x)\)</span> spans.

I think I'm going to need some help from Tim and Matti because this involved the Moodle filtering system.

**Comment by samfearn** [comment]
But the stripping of the backslashes isn't the only problem is it? Because if I escape them, wouldn't I then expect MathJax to kick in and do my formatting? The escaped version clearly works, but requires the additional $x$ to make MathJax load and look for things to process.

**Comment by sangwinc** [comment]
Yes, exactly.  But I think the references above is where the magic happens.

**Comment by aharjula** [comment]
Well, I think the problem is that Markdown requires one to escape the backslash and the MathJax filter looks for math mode starts that are not escaped. But I am not sure in what order those things happen, does MathJax filter act before Markdown gets rendered or not, if so then there is a big problem? In any case I think the problem is pretty much the same as we had many years ago with AJAX validation if the question text had no LaTeX in it i.e. MathJax did not activate if it saw nothing it identified at the phase it checked for things to do.

Based on the example where you escaped the backslashes and added dollars to the text causing things to work I would guess that it would also work if you added e.g. `\(\)` to the end of the text without escapes to trigger MathJax.

In any case if Markdown is to be used castext will need to do escaping for injected values, which means that the castext logic will need to know about the formatter used for the text.

**Comment by sangwinc** [comment]
Yes Matti,
Currently the docs say that CAS text is HTML into which we embed LaTeX equations, CAS expressions, blocks etc. We could, define it that way and end the matter!
With the new API branch, I'd like us to support markdown, since this is a very comfortable human-authoring format.  If this "bug" (or gentle expansion of the supported features perhaps...) is easy to fix, I'd like to do so and fix the behaviour.  
I hope we can do this without making the castext logic know about the format.  Do you have an example of injected values which will break this?

**Comment by aharjula** [comment]
Chris,
Well this might be difficult in any case as the big question is are we targeting normal Markdown here or the Markdown that Moodle has, as there might be "fixes" and non default options in it already? If we target Moodles one then our friends at the ILIAS side will need to interpret Markdown inside STACK questions as Moodle Markdown.

As if we were to target general Markdown then `{@tan(x)@}` would need to escape the wrapper math-mode as well as the backslash of the `\tan` and we still would need the MathJax filter to identify that those bits are truly LaTeX and not something that we have intentionally escaped away from it. Note that I selected "tan" just to point out that `\t` has a role in other escape systems, luckily we have not met those systems otherwise we would surely have seen it already.

In any case backslash is not the only thing that needs to be considered for escaping if we target Markdown with the intention that the result after Markdown has been processed would be something that would look right for MathJax no matter where it is inside that Markdown document as it would not be enough to just know its Markdown it really matters if it is inside a code block or for example link or block quote. Just look at the things GitHubs variant of Markdown has escapes for, also note that the Markdown you write here has its own modifications so many actually that it is called GFM:
https://enterprise.github.com/downloads/en/markdown-cheatsheet.pdf

For those wanting to see how badly the injection currently works just write normal castext with whatever you want and include a code section (i.e. wrapped with back ticks) with e.g. \`\{@\\sqrt{x}@\}\` somewhere in the text to get something through to MathJax. Note that wrapping that in a span with display hidden might solve quite a lot.

Here is a minimal example castext for those needing it, just set to Markdown format:
```
{@tan(x)@} we would want to see this, but the outtermost backslashes get eaten by the rule `\( \)` => `( )`....

`{@tan(x)@}` this is the correct output but as it inside a code-block MathJax skips it, it does however see it and activates to handle the next one, just remove this line and those escapes from the previous one to see what happens.

\\(\\tan{x}\\) and here we have enough escapes, so many actually that MathJax does not see it.

[[input:ans1]] [[validation:ans1]]
```

**Comment by aharjula** [comment]
Also note that the castext parser itself sees the difference between `\\(` and `\(` which leads to even more trouble as it then does not know whether it is in math mode or not.

For a fun example of escaping things based on context try out writing the cube root of something using the wrong parenthesis i.e. `\sqrt[3](x)` and see what happens.

**Comment by ruedigerkupper** [comment]
I confirm that this issue still exists in Moodle 3.9 as described by @samfearn. That behavior is very irritating and clearly a bug. It has been described here, too:
https://moodle.org/mod/forum/discuss.php?d=390891
Is there any progress solving the issue?

**Comment by aharjula** [comment]
I doubt anyone has done anything as this issue has quite a few things that might not be clear:

1. We would need to make STACK inject values as escaped for things to work with MD, however do we also want to output unescaped for those that want to output code fragments? Do we need more injection styles beyond `{@x@}` and `{#x#}`, do both of those get escaped?
2. Do we only escape injected values in the CASText?
3. Should we do the escaping for the whole CASText if it is defined to be MD? i.e. turn `\(` => `\\(` and let the author ignore escaping themselves? In that case we would need to have full LaTeX parsing capabilities to also escape raw LaTeX in the document.

My view is that 1. we do not add new styles of injections, all `{@LaTeX@}`-injections would be escaped when in MD-mode and if that is not enough in some special case people would need to use something else than MD in those situations. 2. only escape injected values, we cannot try to start interpreting the context around the injections. 3. no if you write MD you need to deal with the LaTeX mode activation yourself. No form of mixed-format support will appear and math-mode detection will require escapes if in MD-format. If and as we are likely going to start supportting adaptation of the injection escaping, we will need to define some boundaries and as we cannot start to understand what all happens outside the injection in all the formats we really need to limit ourselves to the injection itself, so lets draw a clear boundary there.

CASText is probably going to be replaced this year with a new implementation so these things may happen there, but for now people really should describe what they want from this and make sure that whatever gets done does not affect people not writing in MD.

The issue of the MathJax filter not triggering without additional configuration or placement of specific hints is not a STACK issue and should be taken to the MathJax filter people, they could check the 'originalformat' in the context they receive for their filter and act accordingly if they see MD. As to STACK not processing its output for a given format, it is not a bug as it does not do that for any format, so a missing feature.

Chris, CASText2 has that extensible `processor` feature, that can be used add new blocks and override old ones, it can be used to do this but it will mean that there will need to be quite a lot of slightly duplicate bits of code to handle this.

**Comment by f0k** [comment]
Chiming in as an affected user. I wanted to write questions in GIFT format with the `[markdown]` specifier, because some things are easier to type in Markdown (e.g., tables), and cannot get LaTeX to work. The problem can also be reproduced elsewhere in Moodle: when I switch the text editor to "unformatted text" and edit some text as Markdown, I also cannot get LaTeX equations to render.

FWIW, since @aharjula asked, I expected Markdown to be rendered to HTML and then processed further as if I had edited text in HTML format. So I expected `\\(x = \\frac{1}{2}\\)` to be rendered as `\(x = \frac{1}{2}\)` and then processed by MathJax.

Just to be sure, did I understand correctly that there is currently no workaround? Adding `$x$` or `\(x\)` doesn't work for me. Or is this all a matter of how Moodle is configured?

**Comment by aharjula** [comment]
Well, as Chris stated there are [places](https://github.com/maths/moodle-qtype_stack/blob/9d0fc6ac98185c9cdc721400f6bbe74325397752/renderer.php#L53) in STACK that need some work to deal with `\\(` or `\\\(` vs. `\(`, and that particular thing breaks any current escaping. So that is where the problem lies.

While I have not gotten there yet the work for supporting injections so that one can get values from Maxima correctly escaped has gotten quite far. So far, that writing `{@sqrt(x)@}` => `\\\(\\sqrt\{x\}\\\)` which would be suitable for Markdown to unescape back to reasonable content for MathJax. The value injection does not currently do escaping (though I expect it will also need to do it) and the logic for injecting to JavaScript contexts like the `[[jsxgraph]]`-block will turn that escaping off, this should provide easier tools for those dealing with inlined graphs and scripts. There will also be a `str_to_md`-function doing escaping if one needs to generate specific fragments at CAS-level. These should be coming to testing during the summer.

However, at this point, even before we start fixing the logic causing the problem itself I really want to know if people are going to write with the ambiguous `\\(` or `\\[` math-mode wrapping or are they going to use the proper markdown escaping `\\\(` and `\\\[` i.e. note that `[` and `(` are special in markdown? For STACK to inject values it needs to know whether it is injecting into math-mode or not, if not then it generates its own wrappers but for it to detect if it is in math-mode it really needs to know that one is not escaping out of math-mode, and in a situation where we have those double or triple slashes it is quite difficult to know what was meant unless one looks even further back in the string, even all the way to start if the string is just backslashes...

_There is no real workaround at the current time, injection of values does not work beyond numbers and triggering MathJax filtering requires special actions._

For now, someone needs to figure out what [these](https://github.com/maths/moodle-qtype_stack/tree/master/stack/mathsoutput) do and maybe do some refactoring and simplification there. At some level, much of their work could be pushed down to CASText2 which would have better contextual awareness.

**Comment by f0k** [comment]
> Well, as Chris stated there are [places](https://github.com/maths/moodle-qtype_stack/blob/9d0fc6ac98185c9cdc721400f6bbe74325397752/renderer.php#L53) in STACK

Just to be sure we are talking of the same thing, is my issue also related to STACK? When I add a topic directly to a course page and edit it as markdown (by using the "plain text" editor and selecting "Markdown format" below the input field), I have the same effect.

> While I have not gotten there yet the work for supporting injections so that one can get values from Maxima correctly escaped has gotten quite far.

Ok, we're not exactly talking of the same thing, I haven't heard of Maxima. If I get this correctly, this is another step on top -- once Moodle supports LaTeX in Markdown, you would like to produce Markdown from some other input. I'm just interested in writing Markdown myself.

> However, at this point, even before we start fixing the logic causing the problem itself I really want to know if people are going to write with the ambiguous `\\(` or `\\[` math-mode wrapping or are they going to use the proper markdown escaping `\\\(` and `\\\[` i.e. note that `[` and `(` are special in markdown?

Oh. I did not think about that. The square bracket `[` would only be a problem if it could be confused with a link, right? And the parenthesis `(` as well? I'd use `\\[` only with a carriage return following, which would not allow it to be confused.
But: In my mental model, anything within the equation that happens to follow a `[foo](bar)` pattern would be parsed as markdown before MathJax kicks in -- so I'd have to take care to escape that, in addition to all the backslashes. That'd make it quite cumbersome to use.

However, I don't know if my mental model represents what's going on in Moodle. I assumed that Markdown -> HTML would happen on the server, and from then on it would be treated the same as if I had entered a HTML snippet. In that case, it wouldn't matter if I used `\\(` or `\\\(`, as it would produce the same HTML. If Markdown is processed completely separately, then as a user, I'd also be fine if it used another syntax for LaTeX altogether. GFM's ``$`\frac{1}{2}`$`` makes pretty clear that there are no additional escape rules to be followed for the LaTeX code, and the same for GFM's:
````
```math
\frac{1}{2}
```
````

**Comment by aharjula** [comment]
This issue was probably more about the escaping of the fragments generated through injection (normally they would come out as unescaped LaTeX, also they might be wrapped with wrong math-mode wrappers) than the general trouble with the Moodle format+filter-stack. Naturally, if the Moodle format+filter-stack does not work it is a bit pointless to work in getting the injections to work for STACKs templating language, but we need both to work for things to make sense in STACK where the material is parametric and you want to inject randomly generated mathematical expressions into the text, and use MarkDown...


Things are complex STACK and the normal Moodle filtering differ a bit, the Moodle way is as follows:
 1. Have text, apply the format conversion e.g. MD, HTML, Moodle-auto-format -> HTML
 2. Apply filters like the MathJax loader on that resulting HTML and it should spot the `\(` or whatever from it to detect if it needs to do something. Assuming that the MathJax filter expects that `\(` and not something else.

If things happen so that the filters act only after the MD->HTML conversion has happened and people have properly escaped their MarkDown things should just work (I assume that Moodle does it in that order, but who knows if it does it like that in every place). Overall, I see no reason for that not to work outside of STACK as long as people remember to escape everything in their LaTeX and note it is not GFM. So you need to escape the values inside the LaTeX environment as well.
 
The STACK way has a slight complication as it does a step before that first one which messes things for STACK + MarkDown. That step is a historical remnant of the times when the output needed to be tuned for various other filters and I really would like to see its creators think its role through in the current context.

As to injections of values generated from the computer algebra system in STACK and injected as parameters into MarkDown templates that is where I come into this and for me the key is that we need to know if we are injecting into math-mode or not and to escape things very differently if we are working with MarkDown. As long as we do not escape for MarkDown things are unlikely to work and as long as we do not detect the math-mode due to not expecting it to be escaped due to being in MarkDown things are messy.

As to `\\[` followed by line change, that just isn't enough experience states that there are plenty of people writing the content directly after the square bracket so I am thinking that STACK will require that things are perfectly escaped, no contextual leniency. We just cannot look at the context. Naturally, that is just STACK whatever others do is their problem. STACK will apply unambiguous escaping on the injected values so expect to see [the full list](https://www.markdownguide.org/basic-syntax/#escaping-characters) of specials to be escaped and in such a way that they will work in tables as well.

**Comment by f0k** [comment]
> This issue was probably more about the escaping of the fragments generated through injection [...] than the general trouble with the Moodle format+filter-stack.

Yes, this dawned on me a bit too late, I was so happy to have found this issue that I misread its purpose.

> Things are complex STACK and the normal Moodle filtering differ a bit, the Moodle way is as follows:
> 
>     1. Have text, apply the format conversion e.g. MD, HTML, Moodle-auto-format -> HTML
> 
>     2. Apply filters like the MathJax loader on that resulting HTML and it should spot the `\(` or whatever from it to detect if it needs to do something. Assuming that the MathJax filter expects that `\(` and not something else.

Just to clear this up, this was my expectation as well, but based on my experience, the order is different:

1. Have text in whatever format, spot the `\(` to know whether to use MathJax or not
2. Apply the format conversion -> HTML
3. Apply MathJax if it was detected in step 1

That's why the `<div style="display:none">\(\)</div>` in the beginning of the Markdown text was required. In case you run into that later, you can auto-generate this along with the escaped Markdown. :crossed_fingers:

**Comment by aharjula** [comment]
Well, [the original issue](https://moodle.org/mod/forum/discuss.php?d=390891#p1578656) on the Moodle site gives some examples of various plugins where things happen in different orders so the trouble will appear in various Moodle plugins in different ways. Naturally, STACK won't be fixing those other places but we will aim to first convert to HTML and then filter with MathJax and other filters so that those do not need to know about Markdown. That order just seems so much more natural so we'll try to make it so in STACK. But as said it is currently broken.

So this Moodle plugin is now aware of the issue, if you are using other plugins you'll need to make them aware of it.
